### DISCUSSION TITLE
Is it possible to add the option to define new services?

### DISCUSSION THREAD

**Discussion by manutenfruits** [zachowj/node-red-contrib-home-assistant-websocket#460] [Feature Request]
I have created a few rudimentary scrapers by creating entities using this plugin, they work great!

However, I'd like to be able to define a few actions on Home Assistant so I can trigger a refresh, dismiss an update, things like that. I understand I am able to do that using a boolean switch (like a helper), but that feels hacky, compared to the potential this plugin brings, defining entities etc.

Would it be possible to define services like `nodered.my_service` or even better to call actions within the defined entity itself (not sure how that'd work, I'm not that familiar with internal Home Assistant models).

Thanks!

**Comment by zachowj** [comment]
If I understand correctly, you can use the existing service `node.trigger` to pass any information you want to an exposed event node or switch entity node.

I realized looking for information on the service it doesn't exist anywhere other than the service docs.

![image](https://user-images.githubusercontent.com/37859597/130361771-2e65b559-fbe2-4bd9-ae64-1f19a9f0c11d.png)

![image](https://user-images.githubusercontent.com/37859597/130362020-a8622da4-9e69-41cf-ac0a-aa87f556cec2.png)

```
[{"id":"be20d11bb705559b","type":"ha-entity","z":"ffbd7f06.4a014","name":"","server":"","version":1,"debugenabled":false,"outputs":2,"entityType":"switch","config":[{"property":"name","value":"testservice"},{"property":"device_class","value":""},{"property":"icon","value":""},{"property":"unit_of_measurement","value":""}],"state":"payload","stateType":"msg","attributes":[],"resend":true,"outputLocation":"","outputLocationType":"none","inputOverride":"allow","outputOnStateChange":false,"outputPayload":"$entity().state ? \"on\": \"off\"","outputPayloadType":"jsonata","x":166,"y":2688,"wires":[["704ec7ea8f9d115f"],[]]},{"id":"376607919b165507","type":"debug","z":"ffbd7f06.4a014","name":"refresh","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":508,"y":2688,"wires":[]},{"id":"26c8ecc06813dbd4","type":"inject","z":"ffbd7f06.4a014","name":"refresh","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"data\":{\"payload\":{\"action\":\"refresh\",\"extra_data\":\"whatever\"}}}","payloadType":"json","x":162,"y":2592,"wires":[["91e451784e65e87b"]]},{"id":"91e451784e65e87b","type":"api-call-service","z":"ffbd7f06.4a014","name":"","server":"","version":3,"debugenabled":false,"service_domain":"nodered","service":"trigger","entityId":"switch.testservice","data":"{\t    \"output_path\": true\t}","dataType":"jsonata","mergecontext":"","mustacheAltTags":false,"outputProperties":[],"queue":"none","x":404,"y":2640,"wires":[[]]},{"id":"eb92045061591064","type":"inject","z":"ffbd7f06.4a014","name":"dismiss","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"data\":{\"payload\":{\"action\":\"dismiss\", \"extra_data\": \"whatever\", \"reason\": \"because\"}}}","payloadType":"json","x":162,"y":2640,"wires":[["91e451784e65e87b"]]},{"id":"704ec7ea8f9d115f","type":"switch","z":"ffbd7f06.4a014","name":"action","property":"payload.action","propertyType":"msg","rules":[{"t":"eq","v":"refresh","vt":"str"},{"t":"eq","v":"dismiss","vt":"str"}],"checkall":"true","repair":false,"outputs":2,"x":354,"y":2688,"wires":[["376607919b165507"],["83bc93afd7207b9f"]]},{"id":"83bc93afd7207b9f","type":"debug","z":"ffbd7f06.4a014","name":"dismiss","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":508,"y":2736,"wires":[]}]
```

**Comment by manutenfruits** [comment]
I had read about `nodered.trigger` from some blog posts, but I wasn't able to find it on my Home Assistant installation. I just checked again and now it does show up, so I can use it. 
The only thing that has changed between then and now is that I have defined a switch entity in nodered to play with. Is it possible that `nodered.trigger` is not available until you create a switch entity? Can you point me to the service docs in case I missed something?

Thank you!

**Reply by zachowj** [reply]
Yes, the service isn't created until NR creates a switch in HA.

**Reply by manutenfruits** [reply]
Okay, so `nodered.trigger` is a service for Nodered defined switches only.
My original question was for a way to define arbitrary services (`nodered.my_service`) so I can trigger flows in nodered directly without using custom defined switches as a workaround.

Assuming that isn't possible today, where can I read more about `nodered.trigger` and what's it supposed to be used for?

**Reply by zachowj** [reply]
> Okay, so `nodered.trigger` is a service for Nodered defined switches only.

Any HA exposed node in NR can be triggered by the `trigger` service call.

> Assuming that isn't possible today, where can I read more about `nodered.trigger` and what's it supposed to be used for?

I wouldn't really even call it documentation just short list parameters that can be used.

https://zachowj.github.io/node-red-contrib-home-assistant-websocket/guide/custom_integration/event-nodes.html

> My original question was for a way to define arbitrary services (`nodered.my_service`) so I can trigger flows in nodered directly without using custom defined switches as a workaround.

It's beyond my knowledge if it would be possible to create dynamic services in HA. I guess I look at `nodered.trigger` as filling that use. In NR, you will have to have some node as a starting point for HA to trigger a flow. Yes, the side effect is a "dummy" switch created in HA but there needs to be something there to tie the two together.

**Reply by manutenfruits** [reply]
> In NR, you will have to have some node as a starting point for HA to trigger a flow.
I was thinking another node like the "entity" node, but called "service" node, and when inserted in NR it would create that service in HA.
